homeassistant/components/motioneye/__init__.py

```diff
@@ -109,13 +149,95 @@ def _add_camera(
 ) -> None:
     """Add a motionEye camera to hass."""
 
-    device_registry.async_get_or_create(
+    def _is_recognized_web_hook(url: str) -> bool:
+        """Determine whether this integration set a web hook."""
+        return f"{WEB_HOOK_SENTINEL_KEY}={WEB_HOOK_SENTINEL_VALUE}" in url
+
+    def _set_webhook(
+        url: str,
+        key_url: str,
+        key_method: str,
+        key_enabled: str,
+        camera: dict[str, Any],
+    ) -> bool:
+        """Set a web hook."""
+        if (
+            entry.options.get(
+                CONF_WEBHOOK_SET_OVERWRITE,
+                DEFAULT_WEBHOOK_SET_OVERWRITE,
+            )
+            or not camera.get(key_url)
+            or _is_recognized_web_hook(camera[key_url])
+        ) and (
+            not camera.get(key_enabled, False)
+            or camera.get(key_method) != KEY_HTTP_METHOD_POST_JSON
+            or camera.get(key_url) != url
+        ):
+            camera[key_enabled] = True
+            camera[key_method] = KEY_HTTP_METHOD_POST_JSON
+            camera[key_url] = url
+            return True
+        return False
+
+    def _build_url(
+        device: dr.DeviceEntry, base: str, event_type: str, keys: list[str]
+    ) -> str:
+        """Build a motionEye webhook URL."""
+
+        return urljoin(
+            base,
+            "?"
+            + urlencode(
+                {
+                    **{k: KEY_WEB_HOOK_CONVERSION_SPECIFIERS[k] for k in sorted(keys)},
+                    WEB_HOOK_SENTINEL_KEY: WEB_HOOK_SENTINEL_VALUE,
+                    ATTR_EVENT_TYPE: event_type,
+                    ATTR_DEVICE_ID: device.id,
+                },
+                safe="%{}",
+            ),
+        )
+
+    device = device_registry.async_get_or_create(
         config_entry_id=entry.entry_id,
         identifiers={device_identifier},
         manufacturer=MOTIONEYE_MANUFACTURER,
         model=MOTIONEYE_MANUFACTURER,
         name=camera[KEY_NAME],
     )
+    if entry.options.get(CONF_WEBHOOK_SET, DEFAULT_WEBHOOK_SET):
+        if (
+            hass.components.cloud.async_active_subscription()
+            and CONF_CLOUDHOOK_URL in entry.data
+        ):
+            url = entry.data[CONF_CLOUDHOOK_URL]
+        else:
+            url = async_generate_url(hass, entry.data[CONF_WEBHOOK_ID])
```

Comment:
So I see that you're generating cloud/external webhooks… but isn't usally MotionEye stored locally? Wouldn't it make sense in that case to try to configure the internal url?

Reply:
If you look at the source for async_generate_url, you can see how it can easily be adapted to use local url. Just copy it here if internal URLs are ok.

Reply:
Yes, I would imagine motionEye would _usually_ (~almost always) be local to the HomeAssistant instance.

Apologies for being dense: Are you suggesting to ditch cloud & external webhooks, and go internal only as a result? Or go internal _and_ cloudhook?

Reply:
Yes, Paulus is suggesting to use full local only. So the implementation of `webhook.async_generate_url` is like this:

```py
@callback
@bind_hass
def async_generate_url(hass, webhook_id):
    """Generate the full URL for a webhook_id."""
    return "{}{}".format(
        get_url(hass, prefer_external=True, allow_cloud=False),
        async_generate_path(webhook_id),
    )
```

So you could generate your own version by e.g.:

```py
@callback
def async_generate_motioneye_webhook(hass, webhook_id):
    """Generate the full URL for a webhook_id."""
    return "{}{}".format(
        get_url(hass, allow_cloud=False),
        async_generate_path(webhook_id),
    )
```

That will prefer a local URL (and fall back to external).

Reply:
I've rolled back the cloudhook support, and implemented the above.